apps/dashboard/src/app/retrieval-agent/agent-dashboard/workflow/workflow.state.ts

```diff
@@ -0,0 +1,145 @@
+import { ComponentRef, signal } from '@angular/core';
+import { ContextAgent, PostprocessAgent, PreprocessAgent } from '@nuclia/core';
+import { ConnectableEntryComponent, NodeDirective } from './basic-elements';
+import { Node, NodeCategory, NodeConfig, NodeType } from './workflow.models';
+
+// TODO sidebar state
+
+// Workflow state
+export const nodes = signal<{ [id: string]: Node }>({});
+export const selectedNode = signal('');
+export const currentOrigin = signal<ConnectableEntryComponent | null>(null);
+export const parentNode = signal<Node | null>(null);
+
+/**
+ * Set selected node and returns it
+ * @param id
+ * @returns
+ */
+export function selectNode(id: string): Node | undefined {
+  const node = getNode(id);
+  if (node) {
+    node.nodeRef.setInput('state', 'selected');
+    selectedNode.set(id);
+  }
+  return node;
+}
+/**
+ * Unselect current selected node if any
+ */
+export function unselectNode() {
+  const node = getNode(selectedNode());
+  if (node) {
+    node.nodeRef.setInput('state', 'default');
+  }
+  selectedNode.set('');
+}
+
+/**
+ * Get the list of all nodes
+ * @returns node list
+ */
+export function getNodes(): Node[] {
+  return Object.values(nodes());
+}
+
+/**
+ * Get specific node
+ * @param id Node identifier
+ * @returns The node corresponding to specified id or undefined
+ */
+export function getNode(id: string): Node | undefined {
+  return nodes()[id];
+}
+
+/**
+ * Add node
+ * @param nodeRef
+ * @param nodeType
+ */
+export function addNode(nodeRef: ComponentRef<NodeDirective>, nodeType: NodeType, nodeCategory: NodeCategory) {
+  nodes.update((items) => ({ ...items, [nodeRef.instance.id]: { nodeRef, nodeType, nodeCategory, children: [] } }));
+}
+
+/**
+ * Update node with agent and config
+ * @param id
+ * @param agent
+ * @param nodeConfig
+ */
+export function updateNode(
+  id: string,
+  agent: PreprocessAgent | ContextAgent | PostprocessAgent,
+  nodeConfig: NodeConfig,
+) {
+  const node = getNode(id);
+  if (node) {
+    nodes.update((items) => ({ ...items, [id]: { ...node, agent, nodeConfig } }));
+    node.nodeRef.setInput('config', nodeConfig);
+  }
```

Comment:
(Just a code style thing)
Here you use `update` to assign the signal value after the logic happened (a bit like we would use `set`).
In my opinion, it is more natural to wrap the full logic in the `update` call, like:
```
nodes.update((items) => {
  const node = items[id];
  if (node) {
    items[id] = {...node, agent, nodeConfig };
    node.nodeRef.setInput('config', nodeConfig);
  }
  return items;
});
```

That would be similar to what we do with the store writers in the widget for example.

But no big deal, the code works like that too, no need to change.

Reply:
I see what you mean. I did it that way because I don't really like the idea of calling a signal (`setInput` is internally a signal) inside another signal’s update…